File: shared/src/commonMain/kotlin/com/cookinlet/belugas/PresenceBanner.kt
package com.cookinlet.belugas

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// "Belugas present" safety banner, first pass: a simple time-decay model, NOT the tidal-based
// heuristic from the original design notes (a future refinement). RED decays to YELLOW after
// [DEFAULT_RED_WINDOW_MS] since the last SELF+geofence-verified sighting in a watched zone;
// YELLOW decays to BLUE after [DEFAULT_YELLOW_WINDOW_MS] since the last sighting of any kind.
// Both are plain constants (not baked into the backend RPC) specifically so they're tunable
// without a migration -- see supabase/migrations/20260829020000_add_beluga_presence_banner.sql.
const val DEFAULT_RED_WINDOW_MS = 2L * 60 * 60 * 1000
const val DEFAULT_YELLOW_WINDOW_MS = 24L * 60 * 60 * 1000

// How close (real distance, not polygon containment) counts as "near" a watched zone for the
// bottom banner's proximity gate -- someone doesn't need to be standing in the river to see
// it. Also client-owned, passed to find_nearby_watched_zone as a parameter, same reasoning as
// the decay windows.
const val DEFAULT_BANNER_PROXIMITY_METERS = 15000.0

// How often to re-fetch the user's location/watched-zone data -- a network round trip, kept
// infrequent. Decay itself (see computeBelugaPresenceStatus) is recomputed locally on a much
// shorter tick so the banner and map shading keep visibly aging between polls.
const val LOCATION_POLL_INTERVAL_MS = 5L * 60 * 1000
const val PRESENCE_DECAY_TICK_INTERVAL_MS = 30L * 1000

// How soon to retry the proximity location fetch specifically after it comes back null (a cold
// GPS fix, most commonly right after launch) rather than waiting out the full poll interval --
// see the retry LaunchedEffect in App.kt for the failure-vs-success branch this feeds.
const val LOCATION_RETRY_INTERVAL_MS = 15L * 1000

// How long since the last SUCCESSFUL presence-state poll (Kenai's get_kenai_presence_state, or
// any other watched zone's get_watched_zone_statuses) before the banner/map admit their data
// might be stale, rather than silently keep showing a last-known RED/YELLOW/BLUE that could be
// hours out of date. 3x the 5-minute poll cadence (LOCATION_POLL_INTERVAL_MS) -- long enough
// that one transient failed poll doesn't flap the indicator on and off, short enough that a
// real sustained outage still surfaces within a quarter hour. Applies uniformly to every watched
// zone, not just Kenai -- see the UNKNOWN case below for the OTHER half of this: staleness is
// about a value that WAS real going out of date, not about never having had one at all.
const val DEFAULT_PRESENCE_STALENESS_THRESHOLD_MS = 3L * LOCATION_POLL_INTERVAL_MS

// UNKNOWN is deliberately its own state, not folded into BLUE -- BLUE is a real, confirmed claim
// ("we checked, nothing recent"); UNKNOWN means the client has never once heard a real answer
// (no successful fetch yet, for this zone, this app run). Collapsing the two used to mean a
// device that had never reached the server rendered a confident "NO RECENT SIGHTINGS" banner and
// blue river shading -- a false all-clear, the one thing get_kenai_presence_state's own
// RED-persistence rule exists to avoid (an overlong RED beats a premature all-clear). A failed or
// not-yet-completed fetch must never resolve to BLUE by default; see App.kt's polling loops
// (both Kenai's and the generic watched-zone one) for how UNKNOWN vs. a real last-known value vs.
// a STALE last-known value are now tracked separately, and SightingsMapScreen for how UNKNOWN
// specifically means "don't draw this zone's shading at all" rather than drawing it in some
// placeholder color.
enum class BelugaPresenceStatus { RED, YELLOW, BLUE, UNKNOWN }

// get_kenai_presence_state's `phase` column, rendered verbatim -- see KenaiPresenceState's own
// comment for why this isn't a local recompute. Any value other than RED/YELLOW fails safe to
// BLUE rather than crashing on an unrecognized string (e.g. if the RPC's phase enum ever grows).
// Never returns UNKNOWN -- that's purely a client-side "no data yet" state, not something the
// phase column itself ever expresses (a successful poll always yields a real phase).
fun belugaPresenceStatusFromKenaiPhase(phase: String): BelugaPresenceStatus = when (phase) {
    "RED" -> BelugaPresenceStatus.RED
    "YELLOW" -> BelugaPresenceStatus.YELLOW
    else -> BelugaPresenceStatus.BLUE
}

// -------------------------------------------------------------------------------------------
// Staleness ESCALATION: a device that's been offline long enough can't just keep showing a
// last-known RED/YELLOW/exempted-BLUE forever with an "(UPDATING...)" suffix (see
// DEFAULT_PRESENCE_STALENESS_THRESHOLD_MS above) -- at some point the claim itself is no
// longer trustworthy and the banner needs to admit it doesn't know, same motivation as UNKNOWN
// existing at all. This only ever moves a status TOWARD UNKNOWN, never invents a fresher one --
// the real value still only ever changes on an actual successful poll.
//
// Cycle boundaries are measured off get_kenai_presence_state's own nextCycleLowEpochMs -- an
// astronomical tide prediction, so it stays valid for the whole outage even though the
// SIGHTING data (what actually makes RED/YELLOW true) can't be refreshed. Only the sighting
// side goes stale; the boundary times don't.
//
// Flat second-boundary offset past the first cycle boundary, rather than predicting the next
// low from the observed current-to-next interval: low-to-low intervals alternate long/short
// (diurnal inequality), so using the preceding interval to predict the next is
// anti-correlated -- it overshoots after a long cycle and undershoots after a short one. A
// flat offset sidesteps that. 12h is the observed MINIMUM low-to-low interval at station 3503
// (real sample: 13:24, 12:08, 13:29, 12:14, 12:45, 12:13, 12:43) -- deliberately the minimum,
// not the average, because the two error directions aren't symmetric: escalating a cycle early
// just shows "STATUS UNKNOWN" slightly sooner, while escalating late means showing YELLOW for
// a stretch nobody actually has sighting data for. Do not "improve" this into an average of
// observed intervals -- that trades the safe direction away.
const val KENAI_SECOND_BOUNDARY_OFFSET_MS = 12L * 60 * 60 * 1000

// When nextCycleLowEpochMs itself is null (the tide predictor hasn't computed that far ahead),
// there's no real boundary timestamp to escalate against -- fall back to a flat ceiling
// measured from the last successful fetch instead. 15h comfortably exceeds every observed
// low-to-low interval at station 3503 (12:08-13:29), so by 15h a real cycle boundary is
// guaranteed to have passed even without knowing exactly when.
const val KENAI_NULL_BOUNDARY_CEILING_MS = 15L * 60 * 60 * 1000

// Ceiling for the two BLUE sub-states get_kenai_presence_state's predicate exempts from
// ordinary cycle-boundary escalation ("NOT EXPECTED THIS TIME OF YEAR" and "PREDICTION
// UNAVAILABLE") -- both non-cycle-scoped claims, so they share one constant rather than each
// getting its own. Strictly, an exempted BLUE is blind to a qualifying sighting landing during
// the outage (RED only ever persists one cycle, so correctness alone would argue for the same
// cycle-scoped rule as everything else) -- but out of season a sighting is rare, and
// escalating every ~12h would leave the banner gray most of the winter for anyone not
// constantly online, training people to ignore it. Alert fatigue is the bigger risk for this
// app, so 24h is the deliberate middle: short enough that no device carries a confident
// all-clear across a meaningful stretch of season, long enough the banner stays meaningful
// rather than perpetually "STATUS UNKNOWN" all winter. Same constant PREDICTION_UNAVAILABLE
// already implied -- not a second number to keep in sync.
const val KENAI_EXEMPT_BLUE_CEILING_MS = 24L * 60 * 60 * 1000

// Applied to gate_time_possible_epoch_ms once, at render, in kenaiBannerLabel below --
// kenai_gate_time itself returns the raw geometric formula result with zero margin by design
// (see that function's own header comment: the offset is deliberately the CALLER's job, not
// folded into the formula). The bias runs EARLY, not late: someone who checks the banner and
// waits three extra minutes for nothing has lost three minutes; someone told "not before X" who
// treats that as a green light and misses the entrance because the tide was already at gate
// depth by X has missed the whole point of the banner. Not rounded to 5 or 10 minutes --
// kenai_gate_time is exact to the minute (worst disagreement verified against 18 hand-tabulated
// logbook cycles was 1.4 minutes), and rounding to a coarser grain would throw away precision
// the formula actually has for no benefit.
const val KENAI_GATE_TIME_EARLY_BIAS_MS = 3L * 60 * 1000

// Pure local recompute of get_kenai_presence_state's own SEASON GATE (two disjoint windows,
// America/Anchorage -- see that function's header comment), used only to catch a device that's
// been offline across a season boundary itself and is holding a now-wrong "NOT EXPECTED THIS
// TIME OF YEAR" claim. This is a DIFFERENT failure than plain elapsed-time staleness --
// crossing a calendar boundary isn't something KENAI_EXEMPT_BLUE_CEILING_MS's ceiling alone
// catches (a device could still be well inside the 24h window and yet already be in a new
// season) -- so both run, not either.
//
// FALL_SEASON_START/END and SPRING_SEASON_START/END keep the SAME LITERAL "MM-DD" STRINGS as
// get_kenai_presence_state's own v_in_season -- a reviewer diffing this file against that
// migration should be able to confirm the two match by eye, not just trust the range logic is
// equivalent. Neither window wraps the Dec 31/Jan 1 year turn, so plain string comparison
// within a single calendar year is safe -- see PresenceBannerSeasonGateTest for the exact
// boundary dates this is expected to hold for.
//
// BOUNDARIES: five years of personal sighting records put the earliest fall arrival at Aug 21
// and the latest spring departure at May 7 (each observed 2021-2025). The windows below pad
// roughly a week past both observed extremes in the direction that matters -- so a slightly
// early or slightly late animal still gets a real BLUE/prediction claim instead of "NOT EXPECTED
// THIS TIME OF YEAR" -- not a claim that these are the true biological limits. Aug 15/Dec 31 and
// Mar 15/May 14 are a judgment call, not derived data; revisit them if more seasons of records
// push either observed extreme wider than the current padding.
private const val FALL_SEASON_START = "08-15" // padding; earliest observed fall arrival is Aug 21
private const val FALL_SEASON_END = "12-31"
private const val SPRING_SEASON_START = "03-15"
private const val SPRING_SEASON_END = "05-14" // padding; latest observed spring departure is May 7

fun isKenaiInSeasonLocally(nowMs: Long): Boolean {
    val monthDay = anchorageMonthDay(nowMs)
    return monthDay in FALL_SEASON_START..FALL_SEASON_END ||
        monthDay in SPRING_SEASON_START..SPRING_SEASON_END
}

// Pairs a fetched KenaiPresenceState with the wall-clock time it was fetched at -- the two are
// always set together (see App.kt's polling loop) and effectiveKenaiPresenceStatus needs both,
// so this makes them impossible to split. They used to be two separate nullable vars in
// App.kt; a caller could (and did, briefly) hand the state to the tick loop with no fetch
// time, and the `?:` fallback that covered that gap defaulted to "elapsed time is zero" --
// i.e. never escalate, the unsafe direction for a feature whose whole point is escalating.
data class KenaiPresenceSnapshot(val detail: KenaiPresenceState, val fetchedAtMs: Long)

/**
 * The banner/map's actually-displayed Kenai status: [snapshot]'s server phase, escalated
 * toward UNKNOWN if [nowMs] has drifted too far past [snapshot]'s fetch time (or past a known
 * cycle boundary) for that phase to still be trustworthy. See the constants above for each
 * ceiling's reasoning.
 *
 * Escalation only ever engages once the snapshot itself is stale
 * (DEFAULT_PRESENCE_STALENESS_THRESHOLD_MS since [snapshot]'s fetch time -- the same threshold
 * that drives the banner's "(UPDATING…)" suffix). Below that, the server phase is returned
 * verbatim regardless of where "now" sits relative to a cycle boundary -- boundaries land twice
 * a day for EVERY device, healthy or not, so escalating on elapsed time alone would gray out
 * the banner for the few minutes between a boundary passing and the next poll landing, on every
 * healthy device, right at the tide turn -- precisely when someone's likely on the bank
 * watching. A genuinely offline device clears DEFAULT_PRESENCE_STALENESS_THRESHOLD_MS long
 * before any of these boundaries or ceilings matter, so gating here doesn't blunt what this
 * function exists to catch.
 *
 * - RED persists until the first cycle boundary (matches the server's own persistence rule),
 *   then reads as YELLOW until the second boundary, then UNKNOWN.
 * - A server-returned YELLOW gets the SAME first+second boundary treatment as RED, not a single
 *   boundary -- get_kenai_presence_state's YELLOW is a pure recency claim spanning the 3
 *   completed cycles before whichever cycle is current (see that function's own comment), so a
 *   YELLOW fetched at any point in that span is just as likely to still be a real YELLOW two
 *   boundaries later as a RED is. There's no "genuine vs. just-expired" distinction left to draw
 *   server-side -- every YELLOW now has the same multi-cycle persistence shape.
 * - A real, non-exempt BLUE ("NEXT WINDOW ..." / the "NO RECENT SIGHTINGS" fallback) is the one
 *   status with no persistence mechanism behind it at all -- a plain "nothing recent" claim, so
 *   it stays cycle-scoped to ONE boundary, straight to UNKNOWN. This is the case the whole
 *   feature exists to close: a stale "no recent sightings" during viewing season is the exact
 *   false all-clear get_kenai_presence_state's own RED-persistence rule was designed to avoid.
 * - The two EXEMPTED BLUE sub-states (out-of-season, prediction-unavailable) are the only ones
 *   that don't use the cycle boundary at all -- they escalate to UNKNOWN past
 *   KENAI_EXEMPT_BLUE_CEILING_MS instead, or immediately if the local season recompute now
 *   disagrees with a stored "not expected" claim (still gated on staleness first, same rule).
 */
fun effectiveKenaiPresenceStatus(
    snapshot: KenaiPresenceSnapshot,
    nowMs: Long
): BelugaPresenceStatus {
    val detail = snapshot.detail
    val lastFetchAtMs = snapshot.fetchedAtMs
    val serverStatus = belugaPresenceStatusFromKenaiPhase(detail.phase)

    if (nowMs - lastFetchAtMs <= DEFAULT_PRESENCE_STALENESS_THRESHOLD_MS) {
        return serverStatus
    }

    val firstBoundaryMs = detail.nextCycleLowEpochMs ?: (lastFetchAtMs + KENAI_NULL_BOUNDARY_CEILING_MS)

    if (serverStatus == BelugaPresenceStatus.BLUE) {
        val isExemptOutOfSeason = !detail.inSeason
        val isExemptNoPrediction = detail.inSeason && !detail.predictionAvailable
        if (!isExemptOutOfSeason && !isExemptNoPrediction) {
            return if (nowMs < firstBoundaryMs) BelugaPresenceStatus.BLUE else BelugaPresenceStatus.UNKNOWN
        }
        if (isExemptOutOfSeason && isKenaiInSeasonLocally(nowMs)) {
            return BelugaPresenceStatus.UNKNOWN
        }
        return if (nowMs - lastFetchAtMs > KENAI_EXEMPT_BLUE_CEILING_MS) {
            BelugaPresenceStatus.UNKNOWN
        } else {
            BelugaPresenceStatus.BLUE
        }
    }

    return when (serverStatus) {
        BelugaPresenceStatus.RED -> {
            val secondBoundaryMs = firstBoundaryMs + KENAI_SECOND_BOUNDARY_OFFSET_MS
            when {
                nowMs < firstBoundaryMs -> BelugaPresenceStatus.RED
                nowMs < secondBoundaryMs -> BelugaPresenceStatus.YELLOW
                else -> BelugaPresenceStatus.UNKNOWN
            }
        }
        BelugaPresenceStatus.YELLOW -> {
            val secondBoundaryMs = firstBoundaryMs + KENAI_SECOND_BOUNDARY_OFFSET_MS
            if (nowMs < secondBoundaryMs) BelugaPresenceStatus.YELLOW else BelugaPresenceStatus.UNKNOWN
        }
        // Unreachable -- belugaPresenceStatusFromKenaiPhase never returns BLUE/UNKNOWN here
        // (BLUE handled above, UNKNOWN never produced by it at all). Kept for exhaustiveness.
        else -> serverStatus
    }
}

/**
 * Pure decay computation: RED if a verified sighting landed within [redWindowMs], else YELLOW
 * if any sighting (verified or not) landed within [yellowWindowMs], else BLUE. [status] null
 * means "a successful get_watched_zone_statuses fetch happened, but had no row for this
 * particular zone" (should not normally happen for a genuinely banner-watched relevant zone,
 * but handled defensively) -- both fold into BLUE, a real "confirmed no recent sightings"
 * answer. Callers must gate on "has a fetch ever actually succeeded" BEFORE calling this --
 * see App.kt's own hasEverFetchedWatchedZoneStatuses -- a status that's null because there is
 * NO data at all yet is UNKNOWN, not BLUE, and this function is never the place that decides
 * that distinction.
 */
fun computeBelugaPresenceStatus(
    status: WatchedZoneSightingStatus?,
    nowMs: Long,
    redWindowMs: Long = DEFAULT_RED_WINDOW_MS,
    yellowWindowMs: Long = DEFAULT_YELLOW_WINDOW_MS
): BelugaPresenceStatus {
    val verifiedAt = status?.lastVerifiedSightingEpochMs
    val anyAt = status?.lastAnySightingEpochMs
    return when {
        verifiedAt != null && nowMs - verifiedAt <= redWindowMs -> BelugaPresenceStatus.RED
        anyAt != null && nowMs - anyAt <= yellowWindowMs -> BelugaPresenceStatus.YELLOW
        else -> BelugaPresenceStatus.BLUE
    }
}

// Shared between the bottom banner and the map's river-shading FillLayer, so both surfaces
// always agree on what RED/YELLOW/BLUE/UNKNOWN actually look like. UNKNOWN's gray is
// deliberately unlike all three real colors -- SightingsMapScreen doesn't actually draw this
// color today (it skips the zone's shading entirely instead, see that file's own comment), but
// the banner does use it directly, and colorForBelugaPresenceStatus needs to stay total either way.
fun colorForBelugaPresenceStatus(status: BelugaPresenceStatus): Color = when (status) {
    BelugaPresenceStatus.RED -> Color(0xFFC62828)
    BelugaPresenceStatus.YELLOW -> Color(0xFFF9A825)
    BelugaPresenceStatus.BLUE -> Color(0xFF0277BD)
    BelugaPresenceStatus.UNKNOWN -> Color(0xFF616161)
}

// Kenai-specific label text, used only when kenaiDetail is non-null (see BelugaPresenceBanner
// below). RED/YELLOW stay identical to the generic labels -- the tide-cycle machinery changes
// WHEN these fire, not what they say once they have (except RED's own "· CHECK MAP" suffix --
// see that case's own comment). BLUE is the one that actually differs: it's
// not a single state here but three (see get_kenai_presence_state's own header comment) --
// out-of-season, in-season-but-no-usable-gate-time, and the real "here's when to expect them"
// case, each needing different text rather than one generic "no recent sightings."
//
// The third case renders gate_time_possible_epoch_ms (0.3m) only -- gate_time_likely_epoch_ms
// (0.8m) stays in the model, unused here, pending a future surface for it. Branches on the
// gate-time field itself (not detail.predictionAvailable) so Kotlin can smart-cast it non-null
// in the branch body below -- the two are equivalent by the server's own definition
// (20260909000000_return_gate_times_from_get_kenai_presence_state.sql), so this is purely to
// avoid a !! rather than a different claim.
private fun kenaiBannerLabel(status: BelugaPresenceStatus, detail: KenaiPresenceState, zoneSuffix: String): String =
    when (status) {
        // RED means a qualifying sighting actually landed this cycle -- the map has something
        // worth looking at right now, so it points there directly. YELLOW deliberately does NOT
        // get the same suffix: it's a recency caution (nobody's seen them for a cycle or more),
        // not a live sighting, and sending someone to the map for something hours old is weaker
        // advice that would dilute "CHECK MAP" as a signal for when RED actually needs it.
        BelugaPresenceStatus.RED -> "BELUGAS PRESENT · CHECK MAP$zoneSuffix"
        BelugaPresenceStatus.YELLOW -> "POSSIBLE ACTIVITY$zoneSuffix"
        BelugaPresenceStatus.BLUE -> when {
            !detail.inSeason -> "NOT EXPECTED THIS TIME OF YEAR$zoneSuffix"
            detail.gateTimePossibleEpochMs == null -> "TIDE DATA UNAVAILABLE$zoneSuffix"
            else -> {
                val biasedGateTimeMs = detail.gateTimePossibleEpochMs - KENAI_GATE_TIME_EARLY_BIAS_MS
                "NOT EXPECTED IN THE RIVER BEFORE ${formatTime12Hour(biasedGateTimeMs)}$zoneSuffix"
            }
        }
        // Unreachable in practice -- BelugaPresenceBanner intercepts UNKNOWN before ever calling
        // this function, since kenaiDetail is null whenever status is UNKNOWN by construction
        // (both come from kenaiPresenceState being null). Kept for when-exhaustiveness.
        BelugaPresenceStatus.UNKNOWN -> "STATUS UNKNOWN$zoneSuffix"
    }

/**
 * Bottom bar reflecting [status] for [zoneName]. The caller decides whether to mount this at
 * all -- it's only shown when the user is subscribed to a watched zone or within
 * [DEFAULT_BANNER_PROXIMITY_METERS] of one (App.kt's showPresenceBanner), on top of the
 * screen-based Placement rule. Once mounted, BLUE is a real, shown state, not a hidden one --
 * it never disappears due to its own color, only because neither gate applies anymore.
 *
 * [status] == UNKNOWN renders as its own distinct gray "STATUS UNKNOWN" -- intercepted before
 * either the Kenai or generic label logic below, since neither has anything meaningful to say
 * about a zone with no data yet. [isDataStale] is unrelated to UNKNOWN (the two are mutually
 * exclusive by construction: staleness is measured from the last SUCCESSFUL fetch, which by
 * definition can't exist yet while status is UNKNOWN) -- it flags "the last successful poll is
 * older than [DEFAULT_PRESENCE_STALENESS_THRESHOLD_MS]" as a suffix on a real last-known
 * [status]/[kenaiDetail], never a fallback to a different computation.
 *
 * [kenaiDetail] is non-null only for the Kenai zone (App.kt passes the fetched
 * KenaiPresenceState through) -- every other watched zone renders the original plain label.
 */
@Composable
fun BelugaPresenceBanner(
    status: BelugaPresenceStatus,
    zoneName: String?,
    modifier: Modifier = Modifier,
    kenaiDetail: KenaiPresenceState? = null,
    isDataStale: Boolean = false
) {
    val backgroundColor = colorForBelugaPresenceStatus(status)
    val zoneSuffix = zoneName?.let { " · ${it.uppercase()}" } ?: ""
    val baseLabel = when {
        status == BelugaPresenceStatus.UNKNOWN -> "STATUS UNKNOWN$zoneSuffix"
        kenaiDetail != null -> kenaiBannerLabel(status, kenaiDetail, zoneSuffix)
        else -> when (status) {
            // Same "· CHECK MAP" reasoning as kenaiBannerLabel's RED case above -- RED is a real
            // qualifying sighting this cycle regardless of which watched zone it's for.
            BelugaPresenceStatus.RED -> "BELUGAS PRESENT · CHECK MAP$zoneSuffix"
            BelugaPresenceStatus.YELLOW -> "POSSIBLE ACTIVITY$zoneSuffix"
            BelugaPresenceStatus.BLUE -> "NO RECENT SIGHTINGS$zoneSuffix"
            // Unreachable (caught by the outer branch above); kept for when-exhaustiveness.
            BelugaPresenceStatus.UNKNOWN -> "STATUS UNKNOWN$zoneSuffix"
        }
    }
    val label = if (isDataStale) "$baseLabel (UPDATING…)" else baseLabel

    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(backgroundColor)
            .navigationBarsPadding()
            .padding(horizontal = 16.dp, vertical = 10.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text(
                text = label,
                color = Color.White,
                fontSize = 12.sp,
                fontWeight = FontWeight.Bold,
                letterSpacing = 0.5.sp,
                textAlign = TextAlign.Center
            )
            if (kenaiDetail != null && kenaiDetail.warnings.isNotEmpty()) {
                Text(
                    text = "⚠ " + kenaiDetail.warnings.joinToString(" · "),
                    color = Color.White.copy(alpha = 0.85f),
                    fontSize = 10.sp,
                    textAlign = TextAlign.Center
                )
            }
        }
    }
}

File: webapp/js/presence-banner.js
// Bottom presence-banner carousel -- ports PresenceBanner.kt's BelugaPresenceBanner/
// BelugaPresenceBannerCarousel + App.kt's card-building (relevantWatchedZoneIds, nonKenai
// escalation) directly.
//
// VISIBILITY, confirmed against source before porting (see App.kt's relevantWatchedZoneIds and
// SupabaseApi.getRelevantWatchedZoneIds' containment-aware RPC): a zone's card only exists when
// this device is subscribed to it (server-side containment match, e.g. an "Entire Inlet"
// subscription also covers Kenai) OR within DEFAULT_BANNER_PROXIMITY_METERS (15km) of it --
// unlike the map's shading, which every viewer sees unconditionally. Neither gate present means
// no card, means nothing renders -- an Anchorage user with no Kenai subscription and nowhere
// near Kenai never sees Kenai's banner, even though they'd still see its river shaded on the map.
//
// Hidden entirely while the Report tab (camera/logging) is showing (app.js's switchTab) --
// matches App.kt's showPresenceBanner excluding CAPTURE/PHOTO_LOGGING/MANUAL_LOGGING/
// ACKNOWLEDGEMENT_GATE. Otherwise it's a fixed bottom overlay above Map/List/the menu/About/
// Resources/News Feed/Alerts, same as native drawing it in the same Box as whatever screen is
// current. Native additionally reserves bottom padding on every AppBackground screen equal to
// the banner's own measured height (LocalBottomContentInset) so it never covers their bottommost
// content -- not reproduced here (a fixed ~50px bar has a small footprint, and dynamically
// measuring/reserving that inset for every screen was judged not worth the complexity for this
// pass); flagged as a known simplification, not an oversight.
let presenceBannerCards = [];
let presenceBannerCurrentIndex = 0;
let presenceBannerOrderKey = "";

// Item 45: full-screen overlays the banner (z-index 970) renders above -- if it's tapped while
// one of these is open, that overlay has to be closed too, not just left covering the map
// underneath after the tab switch. At most one is ever open at once in practice (a subpage is
// only ever reached via the menu, which hides itself first -- see main-menu.js).
const PRESENCE_BANNER_OVERLAY_IDS = [
  "main-menu", "resources-page", "news-feed-page", "about-page", "alerts-page", "share-page"
];

// Item 58 REGRESSION FIX: this used to be a manually-toggled `isReportTabActive` flag, flipped
// only by app.js's switchTab -- correct as long as EVERY way in or out of the Report tab went
// through switchTab, which stopped being true the moment main-menu.js's generic .menu-trigger-btn
// listener (CaptureScreen's own in-context "≡ MENU" button, not just the header's) started
// showing the #main-menu OVERLAY on top of the submit tab without ever calling switchTab: the flag
// stayed stuck true, so the banner never came back once the menu -- not Camera -- was what's
// actually on screen, and nothing was left to notice. Computed fresh from the DOM instead: the
// banner is hidden only while the Report tab is the ACTUAL foreground screen, not merely the
// .view underneath something else -- the menu and every other full-screen overlay
// (PRESENCE_BANNER_OVERLAY_IDS) sit on top of whichever tab and are themselves valid
// banner-visible screens regardless of what's hidden under them (matches App.kt's
// showPresenceBanner exclusion: CAPTURE/PHOTO_LOGGING/MANUAL_LOGGING/ACKNOWLEDGEMENT_GATE only).
function isReportScreenShowing() {
  const anyOverlayOpen = PRESENCE_BANNER_OVERLAY_IDS.some((id) => !document.getElementById(id).hidden);
  if (anyOverlayOpen) return false;
  return !document.getElementById("submit-view").hidden;
}

function initPresenceBanner() {
  document.getElementById("presence-banner-main").addEventListener("click", handlePresenceBannerTap);

  onPresenceStateChanged(rebuildPresenceBannerCards);
  rebuildPresenceBannerCards();

  // Rather than chasing down every place that shows/hides the menu or a full-screen page and
  // pairing it with a manual refresh call (the exact kind of pairing that just drifted out of
  // sync above), watch the real `hidden` attribute on every screen isReportScreenShowing reads
  // and re-evaluate straight from live DOM state on ANY change. Structurally can't drift again --
  // there's no separate flag left to forget to flip. Scoped to exactly these elements rather than
  // `document.body` with subtree:true -- this function's OWN writes below (banner/dots/warnings
  // .hidden) live in that same subtree, and a subtree-wide observer would re-trigger itself on
  // every render, forever.
  const bannerVisibilityObserver = new MutationObserver(() => renderPresenceBannerCard());
  [...PRESENCE_BANNER_OVERLAY_IDS, "map-view", "list-view", "submit-view"].forEach((id) => {
    const el = document.getElementById(id);
    if (el) bannerVisibilityObserver.observe(el, { attributes: true, attributeFilter: ["hidden"] });
  });
}

// Item 45 (field suggestion): RED's own label always ends in "CHECK MAP" -- tapping the banner
// while it reads that way jumps straight to the Sightings Map, the same real destination the
// text is already telling the reader to go check. Every OTHER phase (YELLOW/BLUE/UNKNOWN/LOADING)
// keeps the existing tap-to-cycle-cards behavior unchanged; RED takes priority over cycling on
// the (rare) chance both would otherwise apply to the same tap.
function handlePresenceBannerTap() {
  const safeIndex = Math.min(presenceBannerCurrentIndex, presenceBannerCards.length - 1);
  const card = presenceBannerCards[safeIndex];

  if (card && card.status === PRESENCE_RED) {
    const anyOverlayOpen = PRESENCE_BANNER_OVERLAY_IDS.some((id) => !document.getElementById(id).hidden);
    if (getActiveTabName() === "map" && !anyOverlayOpen) {
      return; // Already looking at the map -- nothing to do.
    }
    const previousTab = getActiveTabName();
    PRESENCE_BANNER_OVERLAY_IDS.forEach((id) => {
      document.getElementById(id).hidden = true;
    });
    switchTab("map");
    pushNavLayer("tab:map", () => {
      switchTab(previousTab);
    });
    return;
  }

  if (presenceBannerCards.length < 2) return;
  presenceBannerCurrentIndex = (presenceBannerCurrentIndex + 1) % presenceBannerCards.length;
  renderPresenceBannerCard();
}

function rebuildPresenceBannerCards() {
  const nearbyIds = presenceState.nearbyWatchedZones.map((z) => z.zone_id);
  const subscribedIds = presenceState.subscribedWatchedZoneIds;
  const relevantIds = Array.from(new Set([...nearbyIds, ...subscribedIds]));
  const nowMs = Date.now();

  const cards = relevantIds.map((zoneId) => {
    const nearby = presenceState.nearbyWatchedZones.find((z) => z.zone_id === zoneId);
    const statusRow = presenceState.watchedZoneStatuses.find((s) => s.zone_id === zoneId);
    const zoneSlug = nearby?.zone_slug ?? statusRow?.zone_slug ?? "";
    const zoneName = nearby?.zone_name ?? statusRow?.zone_name ?? null;
    const isKenai = zoneSlug === "kenai";

    // Item 44: isLoading is distinct from a genuinely-UNKNOWN status -- "never fetched yet" reads
    // as LOADING (presenceBannerLabel, presence.js), not the same "STATUS UNKNOWN" grey a real
    // fetch failure or staleness escalation produces. With initPresenceState() now awaited by
    // app.js's own splash gate, a normal launch never actually reaches this branch at all (the
    // first fetch has already landed by the time the banner can even be shown) -- it's a fallback
    // for the residual case (a fetch failure, or a slower-than-usual network), not the common path.
    let status;
    let isLoading = false;
    if (isKenai) {
      if (!presenceState.hasEverFetchedKenaiPresenceState) {
        isLoading = true;
        status = PRESENCE_UNKNOWN;
      } else {
        status = presenceState.kenaiBelugaStatus;
      }
    } else if (!presenceState.hasEverFetchedWatchedZoneStatuses) {
      isLoading = true;
      status = PRESENCE_UNKNOWN;
    } else {
      const flat = computeBelugaPresenceStatus(statusRow, nowMs);
      status = effectivePresenceStatus(flat, presenceState.lastSuccessfulWatchedZoneStatusesFetchAtMs, nowMs);
    }

    return {
      zoneId,
      zoneName,
      zoneSlug,
      status,
      isLoading,
      kenaiDetail: isKenai ? (presenceState.kenaiPresenceSnapshot?.detail ?? null) : null,
      isDataStale: isKenai ? presenceState.isKenaiDataStale : presenceState.isWatchedZoneStatusesStale,
      isSubscribed: subscribedIds.includes(zoneId)
    };
  }).sort(presenceBannerCardComparator);

  // Reset to the top card whenever the ORDER changes (a new RED, a zone dropping off, a
  // subscribed zone overtaking a nearby-only one) -- not on every routine same-ordering refresh,
  // matching BelugaPresenceBannerCarousel's own orderKey behavior exactly.
  const orderKey = cards.map((c) => c.zoneId).join(",");
  if (orderKey !== presenceBannerOrderKey) {
    presenceBannerCurrentIndex = 0;
    presenceBannerOrderKey = orderKey;
  }

  presenceBannerCards = cards;
  renderPresenceBannerCard();
}

function renderPresenceBannerCard() {
  const banner = document.getElementById("presence-banner");

  if (isReportScreenShowing() || presenceBannerCards.length === 0) {
    banner.hidden = true;
    return;
  }
  banner.hidden = false;

  const safeIndex = Math.min(presenceBannerCurrentIndex, presenceBannerCards.length - 1);
  const card = presenceBannerCards[safeIndex];
  const color = colorForBelugaPresenceStatus(card.status);

  // Item 53: YELLOW is now the same true --brand-yellow (#FFFF00) the rest of this app uses --
  // white text/dots (fine against RED/BLUE's own darker fills) read as barely-there against pure
  // yellow, so .on-yellow (style.css) flips them to black, matching the black-on-yellow
  // convention .chip-toggle.active/native's own yellow FilterChips already use. Toggled on
  // #presence-banner itself, not -main, since -dots is -main's own sibling, not a descendant --
  // both need to flip together.
  banner.classList.toggle("on-yellow", card.status === PRESENCE_YELLOW);

  const mainEl = document.getElementById("presence-banner-main");
  mainEl.style.background = color;
  document.getElementById("presence-banner-label").textContent = presenceBannerLabel(card);

  const warningsEl = document.getElementById("presence-banner-warnings");
  const warnings = card.kenaiDetail?.warnings ?? [];
  if (warnings.length > 0) {
    warningsEl.hidden = false;
    warningsEl.textContent = "⚠ " + warnings.join(" · ");
  } else {
    warningsEl.hidden = true;
  }

  // Item 45: RED always navigates on tap (handlePresenceBannerTap), independent of card count --
  // .clickable/the chevron reflect that this card specifically goes somewhere, same affordance
  // idea as the existing multi-card cycle indicator (dots) below.
  const isRed = card.status === PRESENCE_RED;
  document.getElementById("presence-banner-chevron").hidden = !isRed;

  const dotsEl = document.getElementById("presence-banner-dots");
  const showChrome = presenceBannerCards.length >= 2;
  mainEl.classList.toggle("clickable", showChrome || isRed);
  if (showChrome) {
    dotsEl.hidden = false;
    dotsEl.style.background = color;
    dotsEl.innerHTML = "";
    presenceBannerCards.forEach((_, i) => {
      const dot = document.createElement("span");
      dot.className = "presence-banner-dot" + (i === safeIndex ? " active" : "");
      dotsEl.appendChild(dot);
    });
  } else {
    dotsEl.hidden = true;
  }
}
